rcl_yaml_param_parser/test/test_parser.cpp

```diff
@@ -321,7 +361,9 @@ TEST(RclYamlParamParser, test_parse_file_with_bad_allocator) {
     "correct_config.yaml",
     "empty_string.yaml",
     "indented_name_space.yaml",
-    "max_num_params.yaml",
+    "multiple_params.yaml",
+    "multiple_nodes.yaml",
+//    "multiple_nodes_parameters.yaml",  // Comment it because Timeout while running test.
```

Comment:
Using 'colcon test' will be Timeout, but I ran successfully while run `./test_parser` directly.

Reply:
> Using 'colcon test' will be Timeout, but I ran successfully while running `./test_parser` directly.

It's possible to specify the TIMEOUT for tests (the default is 60 seconds) in the `ament_add_gtest` macro.

That being said, I'm concerned that this test is taking longer than 60 seconds.  I've got a few suggestions here:

1.  Can you do a bit of profiling and find out why it is taking so long?  512 nodes or parameters doesn't seem like that much, and thus I wouldn't expect this to take longer than 60 seconds.
1.  You could simplify this a bit by making one file which has > 512 parameters in a single node, and > 512 nodes in a single file.  That's a bit simpler than what you currently have.
1.  You could split these particular tests out of this file and put it in another test.
1.  Since we are removing the maximum limit now, you could reduce the initial number of parameters and nodes to a much smaller number.  In point of fact, we should probably do that anyway, since in most reasonable scenarios, people will only have a few dozen parameters and a similar number of nodes per file.  This will save users memory, and also make it so that the number of parameters/nodes you need to test "overflow" are smaller.  I'll suggest 128 for both the initial number of parameters and the initial number of nodes, but feel free to make different choices.

Reply:
> Can you do a bit of profiling ? 

Yes, of course.

Uncommented `multiple_nodes_parameters.yaml`, and run google-pprof to get [test_parser](https://github.com/iuhilnehc-ynos/share_files/blob/master/test_parser.prof.pdf)

After removing `multiple_*.yaml` from `test_parser.cpp` and spliting the particular tests in new files `test_parser_multiple_nodes.cpp` and `test_parser_multiple_parameters.cpp`, and run google-pprof to get [test_parser_multiple_nodes](https://github.com/iuhilnehc-ynos/share_files/blob/master/test_parser_multiple_nodes.prof.pdf) and [test_parser_multiple_parameters](https://github.com/iuhilnehc-ynos/share_files/blob/master/test_parser_multiple_params.prof.pdf)


> and find out why it is taking so long

From the cpu profiling of [test_parser](https://github.com/iuhilnehc-ynos/share_files/blob/master/test_parser.prof.pdf) google-pprof , we can find the reason is mainly related to two functions, `find_node` to call `node_param_init`, `yaml_parser_parse`. It seems it's not related to the new added `*_reallocate` functions.


> 512 nodes or parameters doesn't seem like that much, and thus I wouldn't expect this to take longer than 60 seconds.

I supposed you knew it's a test case about bad allocator with macro `RCUTILS_FAULT_INJECTION_TEST`.

FYI:
```
[ RUN      ] test_file_parser.multiple_number_parameters
[       OK ] test_file_parser.multiple_number_parameters (2 ms)
[ RUN      ] test_file_parser.multiple_number_nodes
[       OK ] test_file_parser.multiple_number_nodes (4 ms)
[ RUN      ] test_file_parser.multiple_number_nodes_and_parameters
[       OK ] test_file_parser.multiple_number_nodes_and_parameters (4 ms)
```

> You could simplify this a bit by making one file which has > 512 parameters in a single node, and > 512 nodes in a single file. That's a bit simpler than what you currently have.

I had already provided these two files, please refer to "multiple_params.yaml" and "multiple_nodes.yaml".
OK, I'll only use these two yamls and remove the "multiple_nodes_parameters.yaml".

> You could split these particular tests out of this file and put it in another test.

After using 128 as initial value, `colcon test` can be passed, but I'd like to accept this suggestion to split these particular tests.


> I'll suggest 128 for both the initial number of parameters and the initial number of nodes

OK, I do like this number.

Reply:
> From the cpu profiling of [test_parser](https://github.com/iuhilnehc-ynos/share_files/blob/master/test_parser.prof.pdf) google-pprof , we can find the reason is mainly related to two functions, `find_node` to call `node_param_init`, `yaml_parser_parse`. It seems it's not related to the new added `*_reallocate` functions.

This is great work, thanks for taking the time to do this.  After looking through your profiling a bit, one obvious slowdown is that [find_parameter](https://github.com/ros2/rcl/blob/8bcada7b884693644d7ba2e3bd2c22fd9541eba7/rcl_yaml_param_parser/src/parse.c#L806) just iterates straight over the list, in `O(n)` time.  One thing we *could* do is to change this to a hash table with `O(1)` time for lookups.  That would improve the CPU usage quite a bit.  I'm not suggesting we do that in this PR, of course, but I would ask if you could open a follow-up issue to improve the performance there.

> I supposed you knew it's a test case about bad allocator with macro `RCUTILS_FAULT_INJECTION_TEST`.

I didn't know that, thanks for pointing it out.

> After using 128 as initial value, `colcon test` can be passed, but I'd like to accept this suggestion to split these particular tests.

Yeah, I think that makes more sense anyway.  I find it easier to run the tests individually anyway.

Thanks again for all of this work, I'll re-review the code.

Reply:
> One thing we could do is to change this to a hash table with O(1) time for lookups. That would improve the CPU usage quite a bit. I'm not suggesting we do that in this PR, of course, but I would ask if you could open a follow-up issue to improve the performance there.

I'd like to do that.